### PR TITLE
feat: Add the prototype of goblin translator circuit builder

### PR DESCRIPTION
**Pull request by Rumata888**
# Description
Added the prototype of the translator circuit builder with lots of wires and no selectors. For now the prototype only has one gate - the accumulation gate which computes one iteration of the batch polynomial evaluation.

It seems we have automatic format on commit, so some files got automatically formatted when added from master
# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] The branch has been merged with/rebased against the head of its merge target.
- [x] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [x] There are no circuit changes, OR a cryptographer has been assigned for review.
- [x] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [x] If existing code has been modified, such documentation has been added or updated.
- [x] No superfluous `include` directives have been added.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to any issue(s) it resolves.
- [x] I'm happy for the PR to be merged at the reviewer's next convenience.

### CODE REVIEW

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+    /**
+     * @brief Create a single accumulation gate
+     *
+     * @param acc_step
+     */
+    void create_accumulation_gate(const AccumulationInput acc_step)
+    {
+        // The first wires OpQueue/Transcript wires
+        ASSERT(uint256_t(acc_step.op) <= MAX_OPERAND);
+        auto& op_wire = std::get<WireIds::OP>(wires);
```
Comment (ledwards2225) [inline cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp:261]: Any reason these 3 lines dont use something like `insert_pair_into_wire(WireIds::OP, acc_step.op, zero_idx);`?

File: cpp/src/barretenberg/dsl/acir_format/recursion_constraint.test.cpp [hunk 2]
```diff
@@ -1,7 +1,7 @@
-#include "recursion_constraint.hpp"
 #include "acir_format.hpp"
 #include "barretenberg/plonk/proof_system/types/proof.hpp"
 #include "barretenberg/plonk/proof_system/verification_key/verification_key.hpp"
+#include "recursion_constraint.hpp"
```
Comment (ledwards2225) [inline cpp/src/barretenberg/dsl/acir_format/recursion_constraint.test.cpp:4]: Is it clear what all these unrelated changes are about? You said they're a result of something from Adam? Should they be undone?

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+#include "barretenberg/ecc/curves/bn254/fq.hpp"
+#include "barretenberg/proof_system/arithmetization/arithmetization.hpp"
+#include "circuit_builder_base.hpp"
+#include <array>
+#include <cstddef>
+namespace proof_system {
+class GoblinTranslatorCircuitBuilder : CircuitBuilderBase<arithmetization::GoblinTranslator> {
+    // We don't need templating for Goblin
+    using Fr = barretenberg::fr;
+    using Fp = barretenberg::fq;
```
Comment (ledwards2225) [inline cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp]: Just curious about this notation. Is this p = q thing a notation that we use elsewhere? In other words, I'm curious why not `using Fq = barretenberg::fq;`

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+    GoblinTranslatorCircuitBuilder& operator=(GoblinTranslatorCircuitBuilder&& other) noexcept
+    {
+        CircuitBuilderBase::operator=(std::move(other));
+        return *this;
+    };
+    ~GoblinTranslatorCircuitBuilder() override = default;
+
+    /**
+     * @brief The accumulation input structure contains all the necessary values to initalize an accumulation gate as
+     * well as addition values for checking its correctness
```
Comment (ledwards2225) [inline cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp]: additional*

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+     *
+     * @details For example, we don't really nead the prime limbs, but they serve to check the correctness of over
+     * values. We also don't need the values of x's and v's limbs during circuit construction, since they are added to
+     * relations directly, but this allows us to check correctness of the computed accumulator
+     */
+    struct AccumulationInput {
+        // Members necessary for the gate creation
+        Fr op; // Operator
```
Comment (ledwards2225) [inline cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp]: Maybe in the interest of good searchable names we should switch to using `op_code` or something instead of just `op`. Something to consider. I'm currently using `op` too.

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+    void create_poly_gate(const poly_triple&) override{};
+    [[nodiscard]] size_t get_num_constant_gates() const override { return 0; };
+
+    /**
+     * @brief There are so many wires that naming them has no sense, it is easier to access them with enums
+     *
+     */
+    enum WireIds : size_t {
+        OP, // The first 4 wires contain the standard values from the EccQueue wire
+        X_LO_Y_HI,
```
Comment (ledwards2225) [inline cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp:39]: This ENUM highlights the weird choice we made for the first 4 wires. All of your other wires correspond to one "thing", as opposed to the first 4 which have this weird structure based on the way Zac originally defined the Transcript columns. Maybe it makes more sense to do things the way you've done it here, so op in wire 1, x_lo and x_hi in wire 2, etc. Something to consider but would of course take some coordination at this point.

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+                // The result of division goes into the next evaluation (the same as a carry flag would)
+                // So the lowest wide limb is : (∑everything[0:67]⋅everything[0:67] +
+                // 2⁶⁸⋅(∑everything[0:67]⋅everything[68:135]))/ 2¹³⁶
+                //
+                // The high is:
+                // (low_limb + ∑everything[0:67]⋅everything[136:203] + ∑everything[68:135]⋅everything[68:135] +
+                // 2⁶⁸(∑everything[0:67]⋅everything[204:271] + ∑everything[68:135]⋅everything[136:203])) / 2¹³⁶
+                //
+                // We also limit computation on limbs of op, z_1 and z_2, since we know that op has only the lowest limb
+                // and z_1 and z_2 have only the two lowest limbs
```
Comment (ledwards2225) [inline cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp:645]: These comments are very helpful, thanks for adding them. Without any context (maybe this is all very similar to things done in bigfield) these were very hard to parse. Hopefully we will eventually have thorough documentation of how all of this works but I appreciate that that is difficult as things are in flux.

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+         quotient_witnesses[1] * neg_modulus_limbs[0] - remainder_witnesses[1]) *
+            shift_1; // And this covers the limb shifted by 68
+    for (auto& limb : quotient_witnesses) {
+        info("Q: ", limb);
+    }
+    // Treating accumulator as 254-bit constrained value
+    constexpr auto max_limb_size = (uint512_t(1) << NUM_LIMB_BITS) - 1;
+    constexpr auto shift_1_u512 = uint512_t(shift_1);
+    constexpr auto op_max_size = uint512_t(4);
+    constexpr uint512_t lwl_maximum_value = op_max_size + (max_limb_size * max_limb_size) * ((shift_1_u512 * 12) + 6);
```
Comment (ledwards2225) [inline cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp]: Just flagging that its not immediately clear to me what `lwl` stands for / refers to. Low wide limb?

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+    constexpr uint512_t hwl_maximum_value =
+        lwl_maximum_value_constraint + (max_limb_size * max_limb_size) * 16 +
+        (max_limb_size * max_limb_size * 10 + max_limb_size * max_high_limb_size * 10) * shift_1_u512;
+    constexpr uint512_t hwl_maximum_value_constraint =
+        (hwl_maximum_value >> (2 * NUM_LIMB_BITS)).lo +
+        uint256_t(uint64_t((hwl_maximum_value % uint512_t(1) << (2 * NUM_LIMB_BITS)) != 0));
+    constexpr auto hwl_range_consraint_size = hwl_maximum_value_constraint.get_msb() + 1;
+    info(hwl_range_consraint_size);
+    // 4 high combinations = 8 ml*ml + 8 ml*last_ml. 2 low combinations = 2*ml*ml + 2*ml*last_ml
+    Fr high_wide_relation_limb =
```
Comment (ledwards2225) [inline cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp:189]: This is another chunck that's quite hard to parse. Worth adding comments and/or breaking into smaller components for readability but that can wait as I know this is a WiP

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+        modulus_u512;
+    info("Max quotient: ", max_quotient);
+    info("Max quotient range constraint: ", max_quotient.get_msb() + 1);
+
+    auto [remainder_0, remainder_1, remainder_2, remainder_3, remainder_prime] = base_element_to_bigfield(remainder);
+    std::array<Fr, 5> remainder_witnesses = { remainder_0, remainder_1, remainder_2, remainder_3, remainder_prime };
+    auto [quotient_0, quotient_1, quotient_2, quotient_3, quotient_prime] = uint512_t_to_bigfield(quotient);
+    std::array<Fr, 5> quotient_witnesses = { quotient_0, quotient_1, quotient_2, quotient_3, quotient_prime };
+
+    // We will divide by shift_2 instantly in the relation itself, but f
```
Comment (ledwards2225) [inline cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp]: Seems to be an unfinished thought in this comment

**Review by ledwards2225 — APPROVED** [review]
Overall looks great. If this were a finished product I would likely push for more documentation but I know that things are in flux. When I dig into the formulas I can mostly make sense of everything but it seems like you're leaning on some ideas that are difficult to follow for a reader unfamiliar with the inner workings of bigfield (like me). Anyway, my comments are mostly pointing out typos or minor points for discussion. Feel free to merge when you're ready.

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp (hunk 1 above)
Comment (Rumata888) [reply cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp:261]: Can switch to using the zero. Can't use the zero_idx exactly, but should work out the same

File: cpp/src/barretenberg/dsl/acir_format/recursion_constraint.test.cpp (hunk 2 above)
Comment (Rumata888) [reply cpp/src/barretenberg/dsl/acir_format/recursion_constraint.test.cpp:4]: I tried to undo these changes, but it seems that the update happens, when I commit the files. It seems we have an autoformatter for files being committed, so now when I try to roll back this changes, they automatically come back

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp (hunk 3 above)
Comment (Rumata888) [reply cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp]: No, you are right. Thanks for noticing this. Changed it

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp (hunk 6 above)
Comment (Rumata888) [reply cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp:39]: Let's stick with this weirdness for now and come back to it later, when we have everything working together. It should be easier to change at that point.

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp (hunk 7 above)
Comment (Rumata888) [reply cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.hpp:645]: Yes, I've been living and breathing bigfield for a few months, so there is some professional deformation. Will update when we choose the final state of limbs

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp (hunk 8 above)
Comment (Rumata888) [reply cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp]: Changed

File: cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp (hunk 9 above)
Comment (Rumata888) [reply cpp/src/barretenberg/proof_system/circuit_builder/goblin_translator_circuit_builder.test.cpp:189]: This is the same thing as in the check_circuit. Will add comments later
